### ISSUE TITLE
Validate received CAR Files

### ISSUE THREAD

**Issue by vasco-santos** [web3-storage/web3.storage#1036]
@ribasushi found some invalid CARs that we are storing in backups.

We should discuss if we should inspect and validate the blocks when they are uploaded and fail if the CAR is invalid.

For one of the cars in bafybeiahynaxxbthonbdxyyasbitb52knmftegymlkc6cxttdfcwovsz7m ( don't know which, test )
```
Error: mismatch in content integrity
name: bafkreie3ufxjrzjvee2gvq4xqdncmfgmnioxew3wmqdxgd532dve3fl2tm
data: bafkreicozng7awrdjyglxmi6wnxbcpi7reqzsybbcmrj2sbc44afrzzr4y
```
for one in bafybeie5npyt4gn3c2yav3ntrojyxeaxf4fwdin3dsbgrqh745g4trjxmq
```
Error: mismatch in content integrity
name: bafkreicbs2klwgxlq7cvmxyouhduhr2c6m76clz3uexdtfvlgffgvkahm4
data: bafkreigqgdm74bko7stv6i6nwfsxqyqazwy4jpmdz3zpypdmei6khxc73a
```

**Comment by olizilla** [comment]
Currently, if we are unable to pin the full DAG on ipfs-cluster then that upload will not show as pinned in the UI, which is a start, but really, where the api recieves a CAR with 1 or more cid/block mismatches, we should fail loudly and alert the user that the upload should be re-attempted. 

For larger datasets, we could suggest adding to a local ipfs node and using the pinning api so that only the failed/missing blocks are re-attempted... a fancy resume mechanism thanks to the wonders of dags and content-addressing.

We could conceivably also see bitflips when sending data between the api and ipfs-cluster, or to the s3 backups, which need to be considered.

**Comment by ribasushi** [comment]
Regarding bitflips protection in the pipeline, the workflow should be:
1. CF worker starts 2 independent "threads" while the content is streamed in from user:
1.1 md5sum (yes, md5) of the incoming bytes
1.2 check of individual cid+block pairs within the car file itself
2. Once above completes, add the `Content-MD5` header to the PUT request to S3
3. If s3 returns success you know with many-nines certainty that you have a valid, intact car in s3, and you can say 200 to the user

**Comment by olizilla** [comment]
We already iterate all the blocks in the CAR here, so that seems like a good place to verify each multihash
https://github.com/web3-storage/web3.storage/blob/97aa482f6423357e51e7d35101327ced8418aa2e/packages/api/src/car.js#L271

And! we already sha256 the CAR file, so we should send that to amazon
https://github.com/web3-storage/web3.storage/blob/97aa482f6423357e51e7d35101327ced8418aa2e/packages/api/src/car.js#L230-L233
see: https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html#using-additional-checksums

**Comment by ribasushi** [comment]
@olizilla please double check the APIs ( ideally just test it ). My memory says that only MD5 is validated in realtime on PUT, everything else is a "background task"...

**Comment by olizilla** [comment]
oh no
```
Uploading to s3 with ChecksumSHA256: nope
POST /car 200 OK (942.68ms)
```

**Comment by olizilla** [comment]
the docs are a lie, or at best throughly misleading.
> When you upload an object, you can optionally include a precalculated checksum as part of your request. Amazon S3 compares the provided checksum to the checksum that it calculates by using your specified algorithm. If the two values don't match, Amazon S3 reports an error.
– https://docs.aws.amazon.com/AmazonS3/latest/userguide/checking-object-integrity.html
I shall only consult with @ribasushi on matters arcane from now on.

**Comment by ribasushi** [comment]
@olizilla check the md5 works, maybe they broke it too 😭

**Comment by olizilla** [comment]
ya is better
```
Uploading to s3 with ContentMD5: iZdClRi41qZPEUYm5G9LlA==
BadDigest: The Content-MD5 you specified did not match what we received.
    at /Users/oli/Code/web3-storage/web3.storage/packages/api/dist/worker.mjs:188:47203
    at l (/Users/oli/Code/web3-storage/web3.storage/packages/api/dist/worker.mjs:74:9178)
    at Object.next (/Users/oli/Code/web3-storage/web3.storage/packages/api/dist/worker.mjs:74:8482)
    at d (/Users/oli/Code/web3-storage/web3.storage/packages/api/dist/worker.mjs:74:8060)
    at processTicksAndRejections (node:internal/process/task_queues:96:5) {
  Code: 'BadDigest',
  ExpectedDigest: '8997429518b8d6a64f114626e46f4b94',
  CalculatedDigest: 'iZdClRi42qZPEUYm5G9LlA==',
  RequestId: 'BD9KG0QCAR6HSZMB',
  HostId: '1L8OXn2bs16N33t3sm2UgNHoS1cNVWhSDdy0IFve9UGPokpo3M14NNz4uk5nlfoYMSXy36nL2MQ=',
  '$fault': 'client',
  '$metadata': {
    httpStatusCode: 400,
```

**Comment by olizilla** [comment]
not closed until #1069 lands

**Comment by dchoi27** [comment]
Closing since #1069 is merged!
